client/api/src/leaves.rs

```diff
@@ -99,27 +103,52 @@ where
 	}
 
 	/// Update the leaf list on import.
-	/// Returns a displaced leaf if there was one.
-	pub fn import(&mut self, hash: H, number: N, parent_hash: H) -> Option<ImportDisplaced<H, N>> {
-		// avoid underflow for genesis.
-		let displaced = if number != N::zero() {
-			let parent_number = Reverse(number.clone() - N::one());
-			let was_displaced = self.remove_leaf(&parent_number, &parent_hash);
-
-			if was_displaced {
-				Some(ImportDisplaced {
-					new_hash: hash.clone(),
-					displaced: LeafSetItem { hash: parent_hash, number: parent_number },
-				})
-			} else {
-				None
-			}
+	pub fn import(&mut self, hash: H, number: N, parent_hash: H) -> ImportOutcome<H, N> {
+		let number = Reverse(number);
+
+		let removed = if number.0 != N::zero() {
+			let parent_number = Reverse(number.0.clone() - N::one());
+			self.remove_leaf(&parent_number, &parent_hash).then(|| parent_hash)
 		} else {
 			None
 		};
 
-		self.insert_leaf(Reverse(number.clone()), hash.clone());
-		displaced
+		self.insert_leaf(number.clone(), hash.clone());
+
+		ImportOutcome { inserted: LeafSetItem { hash, number }, removed }
+	}
+
+	/// Update the leaf list on removal.
+	///
+	/// Note that the leaves set structure doesn't have the information to decide if the
+	/// leaf we're removing is the last children of the parent. Follows that this method requires
```

Comment:
Can we not just store the parent hash alongside the child hash? Then we can do this internally. Otherwise it sounds really "complicated" from an api pov.